### PR TITLE
SCRUM-4368 Enable skipping of known issues

### PR DESCRIPTION
**Pull request by markquintontulloch**

### CODE REVIEW

File: src/main/java/org/alliancegenome/curation_api/jobs/executors/LoadFileExecutor.java [hunk 1]
```diff
@@ -220,7 +221,11 @@ protected <E extends AuditedObject, T extends BaseDTO> boolean runLoad(BaseUpser
 					// e.printStackTrace();
 					history.incrementFailed();
 					addException(history, e.getData());
-				} catch (Exception e) {
+				} catch (KnownIssueValidationException e) {
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/jobs/executors/LoadFileExecutor.java:224]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineJavaCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line must not have trailing whitespace

File: src/main/java/org/alliancegenome/curation_api/jobs/executors/LoadFileExecutor.java [hunk 2]
```diff
@@ -220,7 +221,11 @@
 					// e.printStackTrace();
 					history.incrementFailed();
 					addException(history, e.getData());
-				} catch (Exception e) {
+				} catch (KnownIssueValidationException e) { 
+					Log.debug(e.getMessage());
+					history.incrementSkipped();
+				}
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/jobs/executors/LoadFileExecutor.java:227]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.blocks.RightCurlyCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally).

**Comment by markquintontulloch** [comment]
Just thought, probably going to break a couple of integration tests

File: src/test/java/org/alliancegenome/curation_api/base/BaseITCase.java [hunk 3]
```diff
@@ -135,39 +135,32 @@ public void checkFailedBulkLoad(String endpoint, String filePath, HashMap<String
...
-			body("history.counts." + countType + ".total", is(nrRecords)).
-			body("history.counts." + countType + ".failed", is(0)).
-			body("history.counts." + countType + ".completed", is(nrRecords));
+	
+	public void checkSuccessfulBulkLoad(String endpoint, String filePath) throws Exception {
+		checkSuccessfulBulkLoad(endpoint, filePath, 1);
 	}
 	
+	public void checkSuccessfulBulkLoad(String endpoint, String filePath, int nrRecords) throws Exception {
+		checkBulkLoadRecordCounts(endpoint, filePath, "Records",nrRecords, 0, nrRecords, 0);
```
Comment (devopsagr) [inline src/test/java/org/alliancegenome/curation_api/base/BaseITCase.java:155]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>',' is not followed by whitespace.

**Review by oblodgett — COMMENTED** [review]
I like the idea of using an exception to capture these. However I am wondering if we could make a generic exception that both ObjectUpdateException and KnownIssueValidationException could inherit from that way the method definitions only need to `throw GenericException` and in our catch blocks we can split out the different ones. That way we don't have to change the code in the future when we want to add more exceptions. I'll have to have a think about this... and if it would work or not.

**Comment by oblodgett** [comment]
I think thats it. We create a `ValidationException`. `ObjectUpdateException extends ValidationException` and `KnownIssueValidationException extends ValidationException` our methods all `throw ValidationException`. `ValidationException extends Exception` and then we should be able to add more as needed.

**Comment by markquintontulloch** [comment]
> I think thats it. We create a `ValidationException`. `ObjectUpdateException extends ValidationException` and `KnownIssueValidationException extends ValidationException` our methods all `throw ValidationException`. `ValidationException extends Exception` and then we should be able to add more as needed.

Makes sense - will update tomorrow

**Comment by markquintontulloch** [comment]
@oblodgett Exception hierarchy updated as suggested

File: src/main/java/org/alliancegenome/curation_api/interfaces/base/BaseUpsertControllerInterface.java [hunk 4]
```diff
@@ -15,5 +16,5 @@ public interface BaseUpsertControllerInterface<E extends AuditedObject, T extend
 	@POST
 	@Path("/upsert")
 	@JsonView(View.FieldsOnly.class)
-	E upsert(T dto) throws ObjectUpdateException;
+	E upsert(T dto) throws ValidationException, KnownIssueValidationException;
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/interfaces/base/BaseUpsertControllerInterface.java]: So this should be able to just throw ValidationException?

Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/interfaces/base/BaseUpsertControllerInterface.java]: yep, missed that one - will update

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/AGMDiseaseAnnotationDTOValidator.java [hunk 5]
```diff
@@ -39,7 +39,7 @@ public class AGMDiseaseAnnotationDTOValidator extends DiseaseAnnotationDTOValida
 	@Inject
 	AlleleService alleleService;
 
-	public AGMDiseaseAnnotation validateAGMDiseaseAnnotationDTO(AGMDiseaseAnnotationDTO dto, BackendBulkDataProvider dataProvider) throws ObjectUpdateException, ObjectValidationException {
+	public AGMDiseaseAnnotation validateAGMDiseaseAnnotationDTO(AGMDiseaseAnnotationDTO dto, BackendBulkDataProvider dataProvider) throws ValidationException, ObjectValidationException {
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/AGMDiseaseAnnotationDTOValidator.java]: Should be able to drop the ObjectValidationException here?

File: src/main/java/org/alliancegenome/curation_api/interfaces/crud/BaseUpsertServiceInterface.java [hunk 6]
```diff
@@ -1,16 +1,17 @@
...
 
 public interface BaseUpsertServiceInterface<E extends AuditedObject, T extends BaseDTO> {
 
-	default E upsert(T dto) throws ObjectUpdateException {
+	default E upsert(T dto) throws ValidationException, KnownIssueValidationException {
 		return upsert(dto, null);
 	}
 	
-	E upsert(T dto, BackendBulkDataProvider dataProvider) throws ObjectUpdateException;
+	E upsert(T dto, BackendBulkDataProvider dataProvider) throws ValidationException, KnownIssueValidationException;
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/interfaces/crud/BaseUpsertServiceInterface.java]: Also here?

File: src/main/java/org/alliancegenome/curation_api/interfaces/crud/BaseUpsertServiceInterface.java [hunk 7]
```diff
@@ -1,16 +1,17 @@
...
-import org.alliancegenome.curation_api.exceptions.ObjectUpdateException;
+import org.alliancegenome.curation_api.exceptions.KnownIssueValidationException;
+import org.alliancegenome.curation_api.exceptions.ValidationException;
 import org.alliancegenome.curation_api.model.entities.base.AuditedObject;
 import org.alliancegenome.curation_api.model.ingest.dto.base.BaseDTO;
 
 public interface BaseUpsertServiceInterface<E extends AuditedObject, T extends BaseDTO> {
 
-	default E upsert(T dto) throws ObjectUpdateException {
+	default E upsert(T dto) throws ValidationException, KnownIssueValidationException {
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/interfaces/crud/BaseUpsertServiceInterface.java]: And here?

**Comment by markquintontulloch** [comment]
@oblodgett I think I've got them all switched to ValidationException only now

**Review by oblodgett — APPROVED** [review]

**Pull request merged**